### PR TITLE
[ROCm] Add bfloat16 support in linear algebra on ROCm

### PR DESCRIPTION
**Pull request by iotamudelta**
This adds support for gemm-style matrix multiplications with data and output in bf16 to PyTorch on ROCm to the backend (i.e., bgemm).

Enable operators depending on bgemm.

With this change, bf16 matrices on ROCm can be multiplied on the GPU.

### CODE REVIEW

**Comment by iotamudelta** [comment]
@gottbrath this is the BLAS bringup for bf16

**Comment by ezyang** [comment]
This patch is described to be ROCm specific, but the contents of the diff suggest to me that it is turning on bfloat16 on regular CUDA as well. What's going on here? 

Additionally, I didn't see any test modifications.

**Comment by iotamudelta** [comment]
@ezyang thanks for looking at it! I don't think we can discriminate between CUDA/ROCm in the Declarations?

Test cases: that's a good point. We've tested with actual scripts but let me see if we can also enable some unit tests here.

**Comment by ezyang** [comment]
I guess what I'm mostly wondering is, does this PR also accidentally add support for CUDA at the same time? Or will the CUDA paths just error.

**Comment by rohithkrn** [comment]
@ezyang cuda paths will just error if bfloat16 type is used.

**Comment by ezyang** [comment]
Thanks. This looks gtg, just needs to resolve merge conflict

**Review by bddppq — CHANGES_REQUESTED** [review]
Could you add some tests? Maybe refer to https://github.com/pytorch/pytorch/pull/27259 to see how to add bfloat16 tests.

File: aten/src/THC/generic/THCTensorMathBlas.cu [hunk 1]
```diff
@@ -252,7 +252,7 @@ void THCTensor_(addr)(THCState *state, THCTensor *r_, THCTensor *t, THCTensor *v
 
     THCTensor_(freeCopyTo)(state, cr, r_);
   }
-#elif defined(THC_REAL_IS_HALF)
+#elif defined(THC_REAL_IS_HALF)  || defined(THC_REAL_IS_BFLOAT16)
```
Comment (izdeby) [inline aten/src/THC/generic/THCTensorMathBlas.cu]: minor: double space

**Comment by izdeby** [comment]
Can you, please, add some description of what and why are these changes for? Also, add tests

**Comment by iotamudelta** [comment]
@bddppq @izdeby tests are incoming.

@izdeby I thought the title was pretty self-explanatory but added more words to the description now. OK?

**Comment by rohithkrn** [comment]
@bddppq @izdeby @ezyang the tests for gemms are under `tensor_op_tests`. Added `bfloat16_precision`(defaults to 1e-5) arg to the argument list and enabled `bfloat16` tests for gemm ops on ROCm

**Review by izdeby — APPROVED** [review]

**Review by bddppq — APPROVED** [review]

**Pull request closed without merging**